add comparison with Orleans

rkuhn opened the pull request.

gabikliot commented:
There is a lot of inaccuracies in the description of Orleans (not sure why people pick to assume the worse when they lack information, it probably in our human nature), but the main one is about messaging guarantees. We guarantee at most once and not at least once, by default:
http://dotnet.github.io/orleans/Runtime-Implementation-Details/Messaging-Delivery-Guarantees.html

gabikliot commented:
Another inaccuracy:
For example: "Grains can only work in either the fully blocked or fully reentrant modes, limiting the user’s choices to a safe one and a fast one". 
Of course this is completely not true. One can pick to queue a continuation (await) on a different scheduler (thread pool for example). This is achieved by the fact that we fully integrate with .NET TPL library and this is natively supported by TPL: Task.StartNew, Task.Run, Task.ContinueWith all provide overloads to specify a different scheduler or synchronization context.

gabikliot commented:
If you are truly interested I can write a detailed bullet by bullet response to the inaccuracies in this summary.

rkuhn commented:
Of course I'm interested! But please consider that I did not make anything up, all information has been extracted from the source I cited above. If the TR is not the authoritative source of information then please send me the real one so I can redo the comparison myself—although it is a shame that I should have wasted many hours on this by trusting that a paper from the team at Microsoft Research is correct in describing their own product.

rkuhn commented:
Sorry about the tone of my previous comment, time is my most previous resource and the idea of wasting it got to me. I am truly interested in understanding Orleans and thereby extending my understanding of Akka, which hopefully explains my passionate reply.

gabikliot commented:
There was one mistake in the paper, about Messaging Delivery Guarantees. All the rest is correct. 
The other inaccuracies I think stem from making assumptions about things that were not described or described partly. Naturally, one cannot describe all details in a short paper. 
The paper, as its name says, is about the Virtual Actor abstraction and its benefits. Its not a full 100% detailed explanation of all we did in project Orleans in 6 years.

Much more details are in our documentation web site: http://dotnet.github.io/orleans/.
I will provide more detailed comments.

rkuhn commented:
Yes, please comment on the individual lines so that discussion happens within the right context. Thanks.

gabikliot commented on ComparisonWithOrleans.md:
```diff
@@ -0,0 +1,203 @@
+
+  * Akka as a toolkit for building distributed systems, offering the full power but also exposing the inherent essential complexity.
+  
+  * Orleans restricts applicability in order to allow seamless use without understanding distributed computing.
```
I think this sentence is very judgmental. Orleans allows non distributed computing experts to write distributed services, but it does not mean that an expert can not find it useful or necessarily limited in applicability or with less performance. An expert can configure/extend it for a wide range of applications. I would have said: "Orleans simplifies distributed computing allowing non experts to write distributed services. At the same time there is enough extension points and flexibility in Orleans to allow even an expert to customize his services in a flexible way.

rkuhn replied:
The intent of the point-by-point comparison is to highlight the differences between both approaches, not to sell one or the other. Therefore since both Orleans and Akka provide extensibility, the interesting question in this particular section is what each solution targets, and reading the TR I came to the conclusion that—bluntly speaking—Akka requires developers to understand distributed computing while Orleans aims at avoiding the necessity for that. This is supported by quotes such as

> To build a correct solution to such problems in the application, the developer must be a distributed systems expert. To avoid these complexities, we built the Orleans programming model and runtime, which raises the level of the actor abstraction.

and

> This level of indirection provides the runtime with the opportunity to solve many hard distributed systems problems that must otherwise be addressed by the developer.

Since I assume that you are one of the co-authors of the paper, you are in a good position to answer the question: what is the primary focus of Orleans? It is completely fine if Orleans also does other things as secondary concerns, but I get the feeling that the primary focus of Orleans and Akka is fundamentally different, so that is what I would like to highlight here.

I’ll flesh out the intro of “different focus” to clarify this intent of this section.

gabikliot replied:
Yes, I am one of the authors. Undoubtedly, **the primary focus of Orleans is to simplify distributed computing and allow not experts to write efficient, scalable and reliable distributed services**. "Guide the developers down a path of best practices" principle is exactly about that.
As a secondary focus we also provide a flexible platform for more expert developers.

gabikliot commented on ComparisonWithOrleans.md:
```diff
@@ -0,0 +1,203 @@
+  * Akka as a toolkit for building distributed systems, offering the full power but also exposing the inherent essential complexity.
+  
+  * Orleans restricts applicability in order to allow seamless use without understanding distributed computing.
+  
```
There is one very important point that needs to be mentioned. From the very early stages of the runtime, at almost any decision we made, we always came from the standpoint of: first, what  should be the default behavior, what would be the most natural and easy for non expert. Second, how an expert can change the default, how it can have more flexibility.We tried not to sacrifice the 2nd.
You will see examples of this principle: simple first, extensibility second, in a lot of comments I make below.

rkuhn replied:
This is a good point, I shall add a section about this: in contrast, Akka’s primary question is “what is the minimal abstraction that we can provide without compromises?” This means that “good default” for us is not driven by what users might expect, but what we think users will find most useful for reasoning about their program once they have understood the abstraction—familiarity is not a goal in itself. I get the impression that Orleans is making a different choice here, too.

gabikliot commented on ComparisonWithOrleans.md:
```diff
@@ -0,0 +1,203 @@
+  
+#### Lifecycle
+
+  * Orleans Grains do not have a lifecycle, can neither be started nor stopped. Consequently they also cannot fail and be restarted, therefore Orleans does not offer tools for software failure handling—the failure handling aspects concentrate on recovering from hardware crashes.
```
Orleans grain does have a lifecycle, it just doesn't allow external caller to control it. When the grain is activated we call OnActivateAsync, where the app code can do app specific initialization, such as read persistent storage. There is no way to explicitly pass arguments to OnActivateAsync, since the caller is not aware of the fact that grain has activation or not.
When we deactivate, we call OnDeactivateAsync, which can do a cleanup, flush persistent state, ...
Upon crash we indeed to do not invoke any application code (there is no "supervisor").

rkuhn replied:
I’m not comfortable with this formulation as it conflates two layers: aren’t you talking about the lifecycle of an _Activation_? The paper is very explicit about the fact that a Grain does not have a lifecycle, but _Activations_ are indeed started and stopped (on demand, not explicitly) and there can even be multiple Activations alive concurrently for the same Grain. I’d prefer to be really precise here, since this is an important point.

gabikliot replied:
Yes, we provide lifecycle of an Activation. Grain as a logical concept exists always and does not have a lifecycle.

gabikliot commented on ComparisonWithOrleans.md:
```diff
@@ -0,0 +1,203 @@
+  
+#### Automatic Creation
+
+  * Orleans Grains are created automatically whenever needed, which implies that initialization activities cannot have persistent effects—otherwise these would be uncontrollable. OTOH this frees the user from having to consider the need for Grain creation. It also means that asking for a Grain you don’t know which implementation you will get.
```
Of course initialization activities **can** have persistent effects. Inside OnActivateAsync one can do what ever code, such as read the persistent state, mutate it and write back. There are absolutely no restrictions onthat. If one uses the build in optional grain persistence, the runtime automatically reads the persistent state before calling OnActivateAsync and injects the read state into grain in memory state. Thus, in the common simple case of using the Orleans persistence, the app code does not have to explicitly read from storage. But it can if it wishes to.

rkuhn replied:
Ah, understood. What I meant was that “fire all the missiles” (or “charge the credit card”) are not sensible things to do within OnActivateAsync since triggering these external actions presumably should be driven by explicitly intent and since the activation lifecycle cannot be explicitly controlled this would be unreliable. The reason why I said this (which I should also include in the text) is that Actors can be used as process managers that perform external activities explicitly and in a controlled fashion based on lifecycle events, and since I want to highlight differences between the two approaches I’d like to mention this (if true).

gabikliot replied:
Yes, OnActivateAsync  is not explicit. If you need an explicit control of "Do X", you would write it as grain method.

gabikliot commented:
@rkuhn, I am replying inline.
At this stage I also wanted to separate clarifying Orleans issues from asking questions about Akka. We do have a lot of questions about Akka, but I will hold all of them and first would try to get clarity about Orleans, and then as a second step get deeper into Akka.
Would that be OK?

gabikliot commented on ComparisonWithOrleans.md:
```diff
@@ -0,0 +1,203 @@
+  
+#### Virtual Actor Space
+
+  * In Orleans each type of Grain corresponds to a practically infinite space of Grain instances that conceptually all exist from the beginning of the universe to its end. The relation to the physical Actors that implement the Grains is explained similar to virtual and physical memory, but this comparison is misleading since the virtual address space of a process is explicitly populated with the  ...
```
The analogy was used in the on demand paging, and separation of virtual addresses vs. physical addresses. In that context we do think this is a very suitable, an even a great analogy. Virtual actors have multiple aspects, not all of them are similar to virtual memory (the eternal existence is not), but on demand instantiation and automatic reclanmantiaon (similar to virtual memory swap out) is.

rkuhn replied:
Okay, will clarify that the analogy applies to that aspect.

gabikliot commented on ComparisonWithOrleans.md:
```diff
@@ -0,0 +1,203 @@
+
+#### Automatic Actor Placement and Load Balancing
+
+  * Orleans Grains are deployed on silos that can span multiple cluster nodes and the user has no influence on their precise placement or load-dependent movement. This means that making use of elastically provisioned resources is fully automatic and built into the model.
```
This is correct, although we do gradually add more complex policies to impact how load balancing is done. For example, we now have a policy to impact grain placement to balance the number of actors per server (https://github.com/dotnet/orleans/blob/master/src/Orleans/Placement/ActivationCountBasedPlacement.cs).
The idea is that the developer will provide the desired policy (with default policy being a good and reasonable one that handles the majority of cases), but will not need to implement the policy itself. The difficulty in implementing the policy is actually quiet significant: distributed load balancing algorithm has to be robust to fluctuations, spikes, thrashing....most of the programmers don't want to deal with those details. 
We do add more and more polices, to provide more of that "customability and extensibility" I mentioned before. Again, "guide the developers down a path of best practices principle".

rkuhn replied:
Sounds very interesting! In any case I think I’ll just add “direct” in front of “influence” and take this point as settled :-)

gabikliot commented on ComparisonWithOrleans.md:
```diff
@@ -0,0 +1,203 @@
+
+#### Virtual Actors
+
+  * Orleans Grains are identified by their nominal type and a 128-bit GUID.
```
Nominal type  and Guid, or long or string, or  a combination of the three. This was added after the paper written.

rkuhn replied:
Check.

gabikliot commented on ComparisonWithOrleans.md:
```diff
@@ -0,0 +1,203 @@
+
+##### Automatic Instantiation
+
+  * An Orleans Grain may at any given time have one or more activations (active instances) depending on deployment mode. This implies that stateful Grains will have to consider that there may be copies running outside of their control that are responsible for the same piece of data or functionality if automatic scale-out is enabled (see below).
```
It is important to note that automatic scale-out is explicitly chosen by application for only a special type of actors, called Stateless Worker. By application design. For regular stateful actors there is no automatic scale-out  and actor has a single instance always when there is no failures or network partitions and may have multiple instance only in short failure cases or network partitions  (but this has to do with AP vs. CP argument, to which we will get below).

rkuhn replied:
Yes, that is what I tried to express. Shall I move “if automatic scale-out is explicitly enabled by the programmer” to the beginning of the second sentence to make this more clear?

gabikliot replied:
you can. But it also says: "responsible for the same piece of data ". Stateless actors are not responsible for data, they logically do not have state. Of course in practice they may have fields in memory and state, but there is no notion of "actor state" for those.

gabikliot commented on ComparisonWithOrleans.md:
```diff
@@ -0,0 +1,203 @@
+
+##### Location Transparency
+
+  * An Orleans Grain does not have knowledge of its physical location which can also transparently change during the runtime of the system. Users of a Grain’s function do not need to know the location of the Grain activation they are talking to. If my understanding is right then Grain handles can also be sent as part of messages.
```
Yes, grain handles (we call them Grain references) can be sent or persistent. They are purely logical concept.

rkuhn replied:
Right, I saw that after having written this sentence and forgot to remove it.

gabikliot commented on ComparisonWithOrleans.md:
```diff
@@ -0,0 +1,203 @@
+  
+  * Akka distinguishes Actor interactions from normal method invocations by requiring the use of the ActorRef’s `!` or `tell` methods in order to send messages. These method return no result value, receiving a reply from the Actor entails providing an ActorRef where the response is sent to.
+
+In this area the differences between Akka and Orleans are very pronounced and mirror the different intent stated up-front: while Orleans aims for convenience and uses syntax that is familiar to OOP practitioners, Akka makes messaging very explicit and requires the user to define message classes. There is one instance of providing a similar model as async–await, namely the interaction between a Pe ...
```
You wrote:
> permitting other invocations to be processed concurrently “during” the `await` call leads to much of the same race conditions and problems that shared-state concurrency has.

I have to disagree. The two concurrency problems are very different. The shared-state concurrency is a low level data race. It is much harder to handle, requires locks (which limit scalability) and does not compose well. As opposite to that interleaving of turns (or reentrancy mode how we call it), is a much simper concurrency model. There are no data races, no locks are needed. The programmer still has to think in terms of concurrent interleaving, which is not trivial and harder than not to think about them at all, but it is easier than low level data races.

The alternative of one way messaging eliminates the interleaving problem, but of course presents the same issue on the higher level: now the programmer needs to think about coordinating his response with reply. So it moves the problem to another place, does not eliminate it.

rkuhn replied:
Ah, this is interesting! While we completely agree that a certain (nasty) class of concurrency problems disappears and another class remains (where Akka’s answer is that explicitly handling it makes code more readable than implicit threading that breaks the correspondence between lexical scope and execution context), the interesting difference is about exception handling. We strongly believe in the Erlang model, distributed try–catch is not only a non-goal for us, it is considered harmful. With a permanent component like a Grain and a multitude of possible clients of this component, how could it possibly make the application more robust to burden all the clients with handling the faults of one component? Their fault recovery attempts or mechanisms will step on each others’ toes and/or need coordination. This is the reason why having exactly one dedicated supervisor for each component simplifies the design and makes it more robust, it concentrates the knowledge of what to do in case of a failure in exactly one place instead of requiring every possible client to share this knowledge.

Since the TR did not mention this aspect (or I may have overlooked it) I shall add a new section detailing this difference.

jthelin replied:
This difference in the fault recover approaches might be one of the most significant "real" differences between Orleans and Akka, but the details and implications might not initially be apparent to people not so familiar with .NET TPL. 

To slightly simplify things a little, I tend to think of the Akka / Erlang "let-it-crash" approach as __vertical__ error propagation where the exception is sent "up" the actor supervision tree until something handles it and does some appropriate recover action.

Orleans on the other hand uses "try-catch" approach [based on the .NET TPL / async await pattern] which is more of a __horizontal__ error propagation - the exception is sent back "along" the call chain until something handles it and takes some appropriate recover action (catch block).

These are both very valid and useful ways to implement error recovery in a complex distributed system, although I do not personally have enough experience with let-it-crash approach to be able to be able to articulate as a researcher when one approach might be considered objectively "better" than the other. 

This is one area where I think both projects have a lot to contribute to an interesting technical discussion, and lots of teaching / learning opportunities on both sides!

At the moment, the individual evaluation of "best" error handling approach seems to be largely driven by which programming language people are most familiar with. 
I do recognize & declare my own bias here by saying that I never want to go back to programming _without_ the async/await functionality in .NET :)
As a former Java programmer, the EA Orbit Project's port of async/await functionality to Java is certainly an interesting initiative I am watching closely. https://github.com/electronicarts/orbit/pull/29

However, it only takes a quick look at the amazing uptime stats for some of Ericsson's systems written in Erlang to know that the let-it-crash approach has lots of merit too!

gabikliot replied:
Vertical vs. horizontal error propagation is a great analogy!

Another reason we believe in horizontal is that you handle the error in the scope of the request, so you know the request details, you know better how to recover. 
Also, a question: does  it mean in Akka that if the actor throws an exception it automatically dies (let it crash) and the supervisor needs to re-create? It means that if the actor serves multiple clients, and it has thrown due to one client request, you are now effectively "punishing" other clients as well, since the actor is dead for all of them now (they mean to re-lookup in the supervisor for the new actor that was re-created, their in memory state may have been lost, ...)

rkuhn replied:
The important refinement that is missing here is that neither horizontal nor vertical is _the solution_, instead we must acknowledge that both are orthogonal concerns and contribute to the solution. For the definitions I’m using please refer to [the glossary of the Reactive Manifesto](http://www.reactivemanifesto.org/glossary#Failure): Errors are made by clients and need to be signaled back to them while Failures render the service unable to perform its function and need vertical help for recovery. We call that supervision, it might be called differently, but the crucial notion is that clients shall not receive Failures. The other notable fact is that stack traces are meaningless in distributed systems, which makes throwing exceptions even less appropriate for Error handling than it was in classical (local) OO programming: the service should reply with a normal value instead that denominates the error.

In any case, can we conclude that I fix the description of shared-state concurrency problems to clarify that low-level data races are removed while high-level message races remain? And I shall also create a new paragraph describing the difference in failure/error handling philosophy.

gabikliot replied:
>: Errors are made by clients and need to be signaled back to them 
>but the crucial notion is that clients shall not receive Failures

So on that point specifically, you would agree that Orleans makes an excellent implementation of that concept: Client in Orleans definitely do not receive failures, due to auto re-instantiation of Virtual actors, but do receive Errors.


Unrelated, just a comment on Reactive manifesto: indeed, it is a nice writeup, but you would probably agree that it is not THE only or even THE best description of principles behind what I call "a good distributed system". So you will have to forgive me if I personally don't consider Reactive Manifesto as my "Bible" for writing distributed systems and do not judge our protocols necessarily via the lenses of  Reactive Manifesto.
What I consider my "Bible" for writing distributed systems are the works by Lamport, Lynch, FLP, and all that stuff.

rkuhn replied:
My link to the reactive manifesto’s glossary was just for convenience in explaining the difference between failures and errors, nothing more ;-)

Back on topic: how does the runtime in Orleans distinguish between errors and failures if both are just modeled by exceptions? Are libraries consciously written to carefully throw one or the other (if there is a distinction)?

gabikliot replied:
If the grain code throws an error on processing application request, it is considered an error and propagated back to the client.
If the server hosting the grain dies (which I call "the service offered by the grain has failed"), the runtime automatically recovers.

I agree that there might be other types of failures or errors which you may categorize as failures and me as errors. This is a rather mute argument - the boundary is not 100% clear, right? So I don't see why this distinction is so super important. At the end we need to look at all possible/interesting errors/failures and see how we handle them. And in that regard, Orleans handles an important and a major part of failures automatically (but maybe not all possible imaginable failures).

rkuhn replied:
Okay, this clarifies it: the distinction between the two is that Akka considers all exceptions emanating from an Actor as failures whereas Orleans considers exceptions as errors. I think this is big enough of a difference to be noted down, I’ll correct the text to reflect this.

gabikliot commented on ComparisonWithOrleans.md:
```diff
@@ -0,0 +1,203 @@
+  
+  * Instead of trying to offer a familiar (synchronous) user experience Akka makes handling asynchronous results explicit. In the case of the `ask`-pattern—which is analogous to how every invocation works in Orleans—the user should transform the Future explicitly or send its value to an Actor by way of the `pipeTo` pattern.
+
+We found that making asynchronous code look synchronous invites many subtle possibilities of making mistakes, either concerning data consistency (i.e. transparently interleaving different invocations by way of their scheduled closures) or liveness (by blocking a Grain from processing any message until some response has been received). The nature of distributed systems is that messages are sent on ...
```
As I said before, this moves the problem of correlating request to response to a different place, presenting now a different difficulty.

gabikliot commented on ComparisonWithOrleans.md:
```diff
@@ -0,0 +1,203 @@
+  
+  * Akka Persistence is based upon event-sourcing, which means that the PersistentActor emits events that represent state changes to be applied. Snapshots may be used as an optimization to shorten recovery time.
+
+In both cases the persistence provider is configurable and users may choose to persist Actor state independently by other means as well. The biggest difference between snapshots and event-sourcing is that the event history of a PersistentActor contains valuable information that is lost when persisting only snapshots: each event has a well-defined meaning in the business domain and thus the event  ...
```
This is a great point and we agree on the usefulness of event source persistence. This is another important extensibility point to Orleans, which we absolutely welcome as part of the 2nd principle I outlined in the beginning.
We are very actively in the process of adding such a support, with the help and in a full coordination with our open source community: https://github.com/dotnet/orleans/issues/343
Expect this to be implemented very soon.

rkuhn replied:
Great!

gabikliot commented on ComparisonWithOrleans.md:
```diff
@@ -0,0 +1,203 @@
+
+#### Overview
+
+Both systems consist of similar core components: Orleans’ Execution corresponds to akka-actor in that it is responsible for the actual execution of Actors, Messaging corresponds to akka-remote in that it manages TCP connections between the cluster nodes, and Hosting corresponds to akka-cluster plus ClusterSharding. One difference is that akka-actor also allows purely local message sends that are  ...
```
Agree. In our view this is an important aspect of Actors: isolation. One can now say that Akka's local messaging, which sacrifices isolation (inlining actor calls means the args are not copied and also the 2nd call happens in the context  of 1st actor and not the 2nd actor), is against the definition of Actors, as per Carl Hewitt definition (since Akka actors do not always provide isolation). I am not saying that. I am totally fine with Akka's choice to tradeoff isolation with performance. Just pointing out that because of that aspect I do NOT start calling Akka actors not actors (like you below call Orleans actors not actors).

I think abstraction are not sacred, and meant to be revisited and updated. You did it in the context of isolation, we did it in the context of creating/instantiation. Lets just agree that both are good extensions to the initial abstraction of the founding fathers and not criticize each other?

rkuhn replied:
I think you may have misunderstood one aspect: when sending a message to another Actor, that invocation is always asynchronous, the message always goes through the recipient’s mailbox (which in fact is the task that is scheduled on the thread pool). The only optimization that Akka does is to allow the user to send messages by reference without intermediate serialization—we do have a configuration option to turn on full serialization also in the local case in order to verify that your program does not in fact rely upon shared state. This means that we very much uphold the Actor Model, we just chose to have the nanny be opt-in while Orleans chose opt-out ;-)

gabikliot replied:
So there are 2 things about isolation: serialization and dispatch/context.
serialization for local calls is opt in Akka and opt out in Orleans. We agree on that.

As for dispatch, I though you wrote above: 
> One difference is that akka-actor also allows purely local message sends that are optimized to local reference passing whereas Orleans invocations are always sent via the Messaging component that consults with the Hosting component as to where the recipient currently resides.

I read from that that in Akka in the local case you inline the call to be just a regular function call by reference. Meaning the called actor will execute in the context of the calling actor, meaning breaking isolation.
If that is not the case, and you go via mailbox and context switch, than first sorry for miss-understanding.
Second,  this is exactly what Orleans is doing as well. In the local case we go via Dispatch module, of course not via Messaging component like you wrote (we make one function call: "is the destination actor already hosted on this machine". This is not considered going via the messaging component). Thus the argument about efficiency in the local case is not valid. We both have to dispatch to a different context.

rkuhn replied:
Okay, will fix.

gabikliot commented on ComparisonWithOrleans.md:
```diff
@@ -0,0 +1,203 @@
+  
+  * Akka uses a sharding approach where the hash space of the logical Actor names is partitioned into shards of equal size and a ClusterSingleton coordinates the placement of these shards onto cluster nodes.
+
+While Orleans’ Hosting has the ability to relocate individual Actors, it must pay for this ability by a hash table that grows proportionally to the number of Grains used. Akka’s hash-based approach in contrast distributes the full location information (on an as-needed basis) because its size is bounded by the configurable number of shards, which allows it to scale further than Orleans in the numb ...
```
This is a small implementation detail. We can also group actors by ids and map the whole group together (e.g., map each 100 actors via hashing into one directory entry). This will allow to overcome any potential directory size limitations (not that we have seen any such practical limitations, RAM is cheap) at the expense of  a bit less flexible placement. And of course having the directory being distributed eliminates the singleton part of the ClusterSingleton (better load spreading, more resilient, not a single point of failure).

rkuhn replied:
Yes, I agree that this is going very much into the implementation details. BTW: the ClusterSingleton only rarely comes into play, in a running application most shards will be already allocated and all interested nodes have cached their locations. You could say that we have a zero*-hop DHT (with the * being “except certain rare cases”). But I shall explicitly add the cost on our side that individual Actor placement is not possible.

gabikliot replied:
Yes, ours is also a *nearly zero hop DHT*. At the base it is one hop - to the directory, but after the initial lookup it is  cached, so it is zero hop. Very similar, right?
And ours is without grouping. Gives more flexibility at the higher memory cost.

gabikliot commented on ComparisonWithOrleans.md:
```diff
@@ -0,0 +1,203 @@
+
+  * Orleans contains code generators for emitting the serialization and deserialization code for all Grain invocations.
+  
+  * Akka’s Serialization subsystem is fully configurable, users can use any library for mediating between objects and ByteString that they like.
```
We recognize the lack of extensibility for Serialization as a limitation and are working on closing this gap: https://github.com/dotnet/orleans/issues/37. Just a matter of priorities.

rkuhn replied:
yup, that pesky 24h-problem, I know it only too well …

gabikliot commented on ComparisonWithOrleans.md:
```diff
@@ -0,0 +1,203 @@
+
+#### Messaging Guarantees
+
+  * Orleans guarantees *at-least-once* semantics by implementing retransmission until acknowledged, but it does not include sequence numbers to ensure per-pair message ordering, meaning that invocations to a Grain that are issued without waiting for the reply of each previous one before sending the next may be received out of order by that Grain.
```
http://dotnet.github.io/orleans/Runtime-Implementation-Details/Messaging-Delivery-Guarantees.html

rkuhn replied:
Okay, thanks for the correction, will update.

gabikliot commented on ComparisonWithOrleans.md:
```diff
@@ -0,0 +1,203 @@
+    
+    While being quite different from Orleans, Akka is also an implementation of the spirit embodied by the Reactive Manifesto.
+
+Besides these differences in the guiding principles of both tools one particular difference deserves further elaboration. The modeling of Actors via strongly typed interfaces and proxy objects is the most prominent difference between Orleans and Akka; the latter includes the same functionality via the `akka.actor.TypedActor` feature, but that has been deprecated for a long time, awaiting its repl ...
```
Weakly typed (untyped) messaging and interface can easily be layered on tip of Orleans. One of our major customers @yevhen did that here: https://github.com/yevhen/Orleankka.
Layering untyped interface on top of a strongly typed one is much easier than the other way around.

Of course Orleans can offer the equivalent of Akka’s `context.become(): use untyped actor and send it a message with "become" content and the actor logic will change its internal state to change how the next  msg is handled, thus" becoming".

And of course Orleans actors can not only work in "fully blocked or fully reentrant modes". One can mix them up very easily. TPL allows to build an easy wrapper to dom thsioe kinds of things. We use it by ourselves.
One can have a reentrant grain and then use [AsyncSerialExecutor](https://github.com/dotnet/orleans/blob/master/src/Orleans/Async/AsyncSerialExecutor.cs) to handle some grain method calls (with awaits inside) as a logical serial non interleaved execution. TPL is very flexible at allowing to build all kinds of concurrency wrappers.

rkuhn replied:
Thanks for the background on TPL, I was not aware of some of the aspects and will clarify the text once I’m clear on the implications.

gabikliot commented on ComparisonWithOrleans.md:
```diff
@@ -0,0 +1,203 @@
+
+This investigation has been sparked by the question of whether Akka.NET shall implement the «Virtual Actor» abstraction that Orleans offers, in addition to the classical Actor Model copied from Akka. After having seen the differences in goals and philosophy I would like to suggest that Orleans use the terms «Grain» and «Silo» to describe the undoubtedly useful abstraction that it provides, but re ...
+
+  * Orleans’ explicit goal is to hide the Actor-inspired implementation details and offer a hassle-free “elastic object” programming model that does not require distributed programming expertise.
```
Distributed Objects, Elastic Objects, Actors, Virtual Actors, Beans, nano-microscopic services.... :-)

We think Virtual Actors is a great name, as **Virtual* very closely explains what the Orleans actors do differently.

rkuhn replied:
Fair enough, but still I feel compelled to point this out :-)

gabikliot replied:
Great!
A very good discussion! I feel I understand Akka much more now. Thank you.

There are still a couple of points that I would want to clarify about Akka. I will ask about them separately.

rkuhn replied:
Thank you! I just pushed updates to those points where I think we reached agreement, feel free to review and ask more. This is indeed a rather useful exercise.

gabikliot commented:
Thank you @rkuhn. Overall looks much more accurate. There are still a couple of places where you did not incorporate my feedback.
1) the optimized local msgs -  I explained iow both our approaches are similar and that "Akka’s lack of indirection **DOES NOT** allow substantial performance optimizations in this regard", compared to Orleans. We do the same.
2) I think you did not stress that Akak breaks grain isolation, by "the default executors do not respect this constraint and will typically run Future continuations concurrently with the Actor that scheduled them."
3) "Grains can only work in either the fully blocked or fully reentrant modes" I explained how this is not correct. Grain can have a mixed mode as well (via TPL primitives). And Grains can also support "becoming".

rkuhn commented:
Yes, I have not yet incorporated everything, I ran into the wall of my time box (this week is rather busy, I’ll come back to this).

rkuhn commented:
Sorry that it took a little longer, I had added a commit that should address the outstanding comments, please review. If everything is good then I’d like to merge.

gabikliot commented:
Looks good. Thanks.

gabikliot commented:
Ronald, I have a question:
@sergeybykov wanted to comment on one of your previous comments:

"The important refinement that is missing here is that neither horizontal nor vertical is the solution, instead we must acknowledge that both are orthogonal concerns and contribute to the solution. For the definitions I’m using please refer to the glossary of the Reactive Manifesto: Errors are made by clients and need to be signaled back to them while Failures render the service unable to perform its function and need vertical help for recovery. We call that supervision, it might be called differently, but the crucial notion is that clients shall not receive Failures. The other notable fact is that stack traces are meaningless in distributed systems, which makes throwing exceptions even less appropriate for Error handling than it was in classical (local) OO programming: the service should reply with a normal value instead that denominates the error.
In any case, can we conclude that I fix the description of shared-state concurrency problems to clarify that low-level data races are removed while high-level message races remain? And I shall also create a new paragraph describing the difference in failure/error handling philosophy."

but we seem not to be able to find that original comment.

So is my comment about "Distributed Systems Bibles" and Reactive Manifesto not being one of them. I actually think my comment is important and expresses something that people need to hear.

What happened to those comments?

rkuhn commented:
Comments are shown as “outdated diff” in the history above, you can expand them and still read them, but since I corrected the text line that they referred to they don’t apply to the current version any longer (in github’s opinion).

One way to avoid this “comment folding” is to make overall comments (as we are doing right now) or comment on individual commits instead of on the PR—this works better if the history is kept, which I intend to do here.

In any case, thanks for lifting these important ones up to this level, they are easily accessible now.

Concerning failures/errors I just added another commit that should clarify the wording and more precisely capture the difference in philosophy between Orleans and Akka. If the current text is okay from your side then I’d like to merge it—we can always discuss and make changes via new issues or PRs on this repository.

gabikliot commented:
Ohh, I see now! Thanks! That was tricky.

Yes, go ahead please. I say lets merge it all and you can just "publish" this whole doc and then we can potentially submit new pull requests to further extend some points. I think it is in a wip branch. Can maybe make it to the main branch. The current version is definitely good enough as version 1.

rkuhn commented:
Thanks for your help, @gabikliot !

sergeybykov commented:
>The other notable fact is that stack traces are meaningless in distributed systems, which makes throwing exceptions even less appropriate for Error handling than it was in classical (local) OO programming: the service should reply with a normal value instead that denominates the error.

@rkuhn It's interesting that you mentioned it. In Orleans, because of the async RPC as the primary mode of interacting with actors and the automatic propagation of errors, stack traces are actually meaningful. Let me illustrate this with an example.

Say you have a web frontend (FE) that upon receiving a REST request makes a call to method X of grain A. As part of executing `A.X()`, grain A makes a call to method Y of grain B. `B.Y()` in its turn makes a call to storage (ST) that throws an exception, e.g. storage is temporarily unavailable. So you have a call chain of `FE -> A.X() -> B.Y() -> ST`.

It is enough to put a try/catch at the FE level to handle such errors AND to have a meaningful call stack of the error. So if you write something like
```csharp
try
{
    var x = await a.X(args);
}
catch(Exception exc)
{
   log(exc);
}
```
then exception exc will contain a chain of inner exceptions and their call stacks: the original exception throw by ST and exceptions re-thrown by B and A with their respective call stacks:

`ExceptionA (A.X()/A.XImpl()/A.CallB())`
     `ExceptionB (B.Y()/B.YInternal()/B.SaveToStorage())`
           `ExceptionST (ST.Foo()/ST.Bar())`

This is the default behavior, with no error handling logic in A or B. Of course, one can put a try/catch within A.X() and/or B.Y() to analyze the error and retry or report it or to alternate execution of the method, e.g. write to a queue in case the primary storage is unavailable.

We believe this is a very important feature of Orleans - distributed and asynchronous propagation of exceptions. It makes reasoning about errors in a distributed app almost comparable to it in a single process app and allows developers to write the minimum amount of code for most common cases, as in the example above - put error handling logic only at the FE level. With pure one-way message passing such simplicity is pretty much impossible to achieve.

rkuhn commented:
@sergeybykov Yes, you’re right: if the service’s model fits request–response well then that gives you a convenient handle on error propagation, as noted in the comparison document. My point is that assuming an exception to be an error (as opposed to a failure) unless stated otherwise can misguide system design, and I certainly do not agree that all error handling logic should be moved up to the FE level. What should be done at that level is user input validation, and if valid inputs lead to errors further down then someone did a mistake internally—which means service failure and not user error.

sergeybykov commented:
@rkuhn I didn't mean that error handling should be done at the FE level. It can definitely be done at any level in the call chain that makes sense for the app. What we see in practice in interactive services though is that it's a rather rare case that a useful error handling can be done at lower levels.

For example, a retry can be attempted by any actor in the call chain. However, in most cases the lower levels cannot know if a retry is desirable by the scenario, which only the top layers in the call chain, FE or not, may know.

I'm not sure I agree with your black and white picture of errors vs. failures. In a distributed system sometimes you simply don't know for sure. A socket error may indicate a failure of the node on the other end or just a temporary network glitch. From the application perspective they are indistinguishable until we learn that this is in fact a failure of the remote node. In the meantime, the app usually cannot wait for such a fact to be established, and has to treat this ambiguous situation as an error.

rkuhn commented:
We should probably meet sometime and discuss this over a beverage of your choice ;-)

One more thought here is that the black/white separation’s feasibility depends on the definitions: it gets pretty clear and simple if you apply HTTP thinking, as in 4xx status codes mean “you did something wrong” (i.e. an Error) while 5xx status codes mean “I did something wrong” (i.e. a Failure). Then it does not matter whether I know that another remote node is down or not—if I cannot provide _my_ service then it is a Failure on my part. This keeps all observations and their reactions nicely local, which is a big plus in distributed systems.

sergeybykov commented:
Definitely. So many questions can be sorted out over a drink. :-) Let me know when you travel to Seattle next time.

domdorn commented:
Are there any updates on this? I just stumbled today over https://github.com/orbit/orbit and two things that would be nice in Akka jump right into my eye:

1.  all actors are apparently persistent and automatically respawn on a different node if the node they are running on crashes
2. they have a nice way of defining an actors protocol..  Something that's still not that nice even with Akka Typed.

domdorn commented:
Oh, I'm sorry, I've just seen that this is about Orleans, but I was looking at "Orbit" .. please ignore my comment.
